datasource/schema/dynamic_attribute.go

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package schema
+
+import (
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/internal/fwschema"
+	"github.com/hashicorp/terraform-plugin-framework/internal/fwschema/fwxschema"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+// Ensure the implementation satisifies the desired interfaces.
+var (
+	_ Attribute                                = DynamicAttribute{}
+	_ fwxschema.AttributeWithDynamicValidators = DynamicAttribute{}
+)
+
+// DynamicAttribute represents a schema attribute that is a dynamic, rather
+// than a single static type. Static types are always preferable over dynamic
+// types in Terraform as practitioners will receive less helpful configuration
+// assistance from validation error diagnostics and editor integrations. When
+// retrieving the value for this attribute, use types.Dynamic as the value type
+// unless the CustomType field is set.
+//
+// TODO: describe some of the rules about dynamic types being determined (literals vs. other schema attributes)
+//
+// TODO: describe how this may be represented in config
+//
+// TODO: describe how dynamic attributes are referenced
```

Comment:
Is it worth saying something about when it makes sense to use a `DynamicAttribute`. Are there a set of criteria that can be applied which would explicitly indicate that a `DynamicAttribute` is warranted?